### ISSUE TITLE
Not able to render Code Coverage results in the "Code Coverage" tab in Azure DevOps

### ISSUE THREAD

**Issue by pradudev** [microsoft/azure-pipelines-tasks#13632]
I have a classic build pipeline in Azure DevOps for that builds and run tests for a .Net Core 3.1 app
I am using self-hosted build agents running Windows Server 2019 OS

Below is the pipeline screenshot
![image](https://user-images.githubusercontent.com/3319134/94642808-9b0fa100-0328-11eb-9b49-54b6b2006742.png)

My tasks:

1. Restore Nuget
2. Build the solution in release mode
3. Install latest vstest.console.exe tool
4. Run tests using above tool and generate .coverage file
5. Install CodeCoverage.exe from Nuget
6. PowerShell script to convert .coverage to .xml
7. Using ReportGenerator task to generate HTML and Cobertura (HTMLInline;HTMLChart;Cobertura)
8. Publish the code coverage results to the server using Cobertura.xml from step 7 and also uploading HTML report folder

**Build runs successfully**
Publish Code Coverage task log
![image](https://user-images.githubusercontent.com/3319134/94643313-dc548080-0329-11eb-8387-833511f99eec.png)

**Build run summary**
![image](https://user-images.githubusercontent.com/3319134/94643354-f1311400-0329-11eb-9cf8-77c89824a003.png)

**Published files by the build**
![image](https://user-images.githubusercontent.com/3319134/94643472-3b19fa00-032a-11eb-872b-b81c1c8eaa6d.png)

**Code Coverage tab**
![image](https://user-images.githubusercontent.com/3319134/94643390-04dc7a80-032a-11eb-9bfa-8131e5f95b2a.png)

**Issue: The code coverage tab does not display the HTML reports**

**Comment by S121Pradhan** [comment]
@pradudev Can you share the .runsettings file ?

**Comment by Caiptain1** [comment]
@pradudev I had the same thing. I had reached out to support for ADO but they weren't very helpful. I basically drove the full process and figured out the cause of this.
### The findings
I came to the conclusion that if the VsTest task uploads the test results, your custom report being uploaded is ignored. When the VsTest task is uploading the test report to ADO, the coverage is an attachment to that and automatically gets uploaded. Since the pipeline already has code coverage registered it doesn't accept anything else. Since ADO doesn't merge the coverage files it kind of makes sense but then you can only publish the coverage once in the whole pipeline.
### The workaround
- Run vstest from a powershell script
- Copy or move the test results file to a different location
- Publish the test results file (since the results have a relative path reference to code coverage, the publish task won't find it and you'll have no code coverage published)
- Convert the .coverage file to HTML report
- Publish code coverage

**Comment by Caiptain1** [comment]
P.S. Microsoft support said they won't do anything about it unless it gets community support on their developer communities. There was an idea raised some time ago to bring back the .coverage conversion directly in ADO: https://developercommunity.visualstudio.com/idea/366142/support-vstest-coverage-code-coverage-build-result.html
I asked my colleagues and some friends to give some support to the idea and it finally passed the threshold and got onto the roadmap for ADO.

**Comment by h0p3zZ** [comment]
I have tried a lot to solve his problem...

So first of all, the problem is the test task of the pipeline (in my case the DotNetCoreCLI@2 task). If this task collects the code coverage it is automatically published with the tests in the .coverage format. For some reason this push of the .coverage file is at the end of the pipeline (or job - this I do not know) and therefore overrides all the previously uploaded files.

### **The Solution**

Use a Coverage Tool like OpenCover and collect the code coverage with that instead of the collection via the test task with CodeCoverage collection on.

Example:
`OpenCover.Console.exe -target:"dotnet.exe" -targetArguments:"test {solution.sln]" -output:"{outputDir}\OpenCover.xml"` 

Afterwards convert the OpenCover file via the ReportGenerator task into Cobertura format.

And after that you can publish it as usual via the PublishCodeCoverageResult task.

You might have to try arround a little but with that solution in mind it will work.

**Comment by elninoisback** [comment]
Your issue is with Step 4. => "Run tests using above tool and generate .coverage file"

If a .coverage file is created and left alone, Azure DevOps will pick it up and use it over anything else you create later. So basically your coverage.xml or any generated HTML report is ignored at the Code Coverage tab.

So instead of using VSTest task and whatever you used beneath that, try adding a .Net Core Test to your pipeline and configure it to spit out "XPlat Code Coverage" using coverlet collector. Make sure NOT TO pass in any .runsettings file via --settings. Before doing any of that, you will need to add the NuGet coverlet.collector for all your test projects.

![image](https://user-images.githubusercontent.com/6582613/130321243-75608c34-0054-4699-90bf-d9fb707d70c0.png)

Then add your task to publish the coverage results. You don't need the report generator, it's now built-in.

![image](https://user-images.githubusercontent.com/6582613/130321252-49d9efc1-2c39-4629-ac34-a4a1cb96ef88.png)

That's all. Hopefully, you will be able to see the generated HTML getting rendered under Code Coverage tab.

![image](https://user-images.githubusercontent.com/6582613/130321259-04714e2c-533f-45ca-991f-1b9a3fe99a51.png)

**Comment by kenneth-westbrook** [comment]
Thank you @elninoisback your guide was VERY helpful!  Got my stuff to publish after following your example.

**Comment by gioce90** [comment]
There is another way to achieve this (the Full coverage on the "Code coverage" tab on the build), without using coverlet.

Now [thanks to](https://github.com/microsoft/vstest/issues/981) https://github.com/microsoft/vstest/issues/981#issuecomment-973994625) is possible to use dotnet test with cobertura format:
```
  - task: DotNetCoreCLI@2
    displayName: 'DotNetCoreCLI Test with report (cobertura format)'
    inputs:
      command: test
      projects: '**/**.Tests.csproj' # if multiple test projects are to be executed, create a dirs.proj and use it here. That will ensure the tests are executed in parallel
      arguments: '--no-restore --collect "Code Coverage" --logger trx --results-directory "TestResults/Coverage/" -- DataCollectionRunSettings.DataCollectors.DataCollector.Configuration.Format=cobertura'
      publishTestResults: true
```
You can also use `publishTestResults: false` instead, followed by `reportgenerator`, `PublishCodeCoverageResults` and `PublishCodeCoverageResults` (with variable `disable.coverage.autogenerate: 'true'`) . It depends of your preference on the reports.

This is good for the full coverage. For the diff coverage on the pull request instead, saddly you have to generate .coverage file. Actually I have two different pipelines (one for the full coverage as showed, the second for the diff coverage)

**Comment by ManikandanMani** [comment]
> * results

Hi @pradudev,
Could you let us know, do you able to convert the .coverage file to xml file, like the below command in powershell task 
### Powershell Task Inline command to convert to xml file 

$coverageFile="$(Common.TestResultsDirectory)\$(build.buildnumber)-$(build.buildid).coverage"
$xmlFile="$(Common.TestResultsDirectory)\$(build.buildnumber)-$(build.buildid).xml"
&"$env:USERPROFILE\.nuget\packages\microsoft.codecoverage\17.0.0\build\netstandard1.0\CodeCoverage\CodeCoverage.exe" analyze /output:$xmlFile  $coverageFile
I used the above command in local it works fine, in the pipelines the coverage file exists and and I can execute the codeCoverage.exe, but I couldn't able to execute the above command altogether since it failed for no reason and I couldn't find why its failing only on the pipelines, have you faced this issue and could you share your workaround for this.
![image](https://user-images.githubusercontent.com/42301172/156914608-77279266-2e3f-494b-9db2-a38491e07ed3.png)

**Comment by fhnaseer** [comment]
@ManikandanMani You can also use dotnet-coverage cli tool to convert coverage files. https://docs.microsoft.com/en-us/dotnet/core/additional-tools/dotnet-coverage#dotnet-coverage-merge
dotnet-coverage merge -f xml sample.coverage

**Comment by ManikandanMani** [comment]
@fhnaseer , Thanks for the info I'll try that.

**Comment by ManikandanMani** [comment]
@fhnaseer, I could able to convert into xml file, but this file is generated under userprofile, also I have an Invalid file issue when I downloaded the coverage file by using the below powershell script, but if I directly download the same coverage file from the codecoverage tab in the build, the convertion works. could you take a look on this.
![InvalidFileWhenDownloadedByPs](https://user-images.githubusercontent.com/42301172/157384563-58385326-0572-4917-9566-0598f9a3be84.png)

![GeneratesOnUserProfile](https://user-images.githubusercontent.com/42301172/157384759-1e00e5f9-7514-43be-87e7-6670d5ecc9f8.png)

In this below picture,the coverage file downloaded are for the same build, not sure about the size difference.

1. The First File starts with 20220304.6...1968 is manually downloaded.
2. The second File with Name 9746 is downloaded by invoking a PS Script.

![SizeDifferenceOfDownloadedcoveragefile](https://user-images.githubusercontent.com/42301172/157384818-c26d4024-37ca-4d53-abe9-c3c4b2339a1e.png)

```
$UriOrga = "https://dev.azure.com/$($OrganizationName)/$($Project)/" 
$uriAccount = $UriOrga + "_apis/test/codecoverage?buildId=9746&flags=1&api-version=6.0-preview.1"
$json=Invoke-RestMethod -Uri $uriAccount -Method get -Headers $AzureDevOpsAuthenicationHeader

$downloadedFile="$env:userprofile\Downloads\9746.coverage"

Invoke-WebRequest $json.psobject.properties["value"].Value.codeCoverageFileUrl -OutFile $downloadedFile
```

**Comment by fhnaseer** [comment]
@ManikandanMani You can also provide output path in command line [-o|--output <output>]. Regarding downloading with ps scritp, I am not sure about it. But it seems like 9746.coverage file is not valid.

**Comment by simonachmueller** [comment]
@nehsharmMS mentioned [here](https://github.com/microsoft/azure-pipelines-tasks/issues/12183#issuecomment-581833577) that MS has new preview option to enable for Code Coverage, does it only require .coverage file? Then coverlet etc. would be not needed.

**Comment by ManikandanMani** [comment]
> @nehsharmMS mentioned [here](https://github.com/microsoft/azure-pipelines-tasks/issues/12183#issuecomment-581833577) that MS has new preview option to enable for Code Coverage, does it only require .coverage file? Then coverlet etc. would be not needed.

@simonachmueller, are you meaning to say that the .coverage file generated in the pipelines would also provides individual coverage metrics for each file covered ?,  btw is this can be achieved in vstest.console.exe [ships with visual studio] or using dotnet test cli? correct me if i am wrong here.

**Comment by ManikandanMani** [comment]
> @ManikandanMani You can also provide output path in command line [-o|--output ]. Regarding downloading with ps scritp, I am not sure about it. But it seems like 9746.coverage file is not valid.

@fhnaseer , I found out the reason why the downloaded file was corrupted ,because I forgot to add the authentication as headers when downloading the coverage file, after adding the headers downloads the coverage file without any issues

Invoke-RestMethod -Uri $json.psobject.properties["value"].Value.codeCoverageFileUrl Get -ContentType "application/json" -Headers @{Authorization=("Basic {0}" -f $AzureDevOpsAuthenicationHeader)} -OutFile $downloadedFile

**Comment by simonachmueller** [comment]
> > @nehsharmMS mentioned [here](https://github.com/microsoft/azure-pipelines-tasks/issues/12183#issuecomment-581833577) that MS has new preview option to enable for Code Coverage, does it only require .coverage file? Then coverlet etc. would be not needed.
> 
> @simonachmueller, are you meaning to say that the .coverage file generated in the pipelines would also provides individual coverage metrics for each file covered ?, btw is this can be achieved in vstest.console.exe [ships with visual studio] or using dotnet test cli? correct me if i am wrong here.

As I understood, there is a preview feature on Azure DevOps to use .coverage file to display code coverage in the build results without needing to generate html results separately.

**Comment by gioce90** [comment]
> @simonachmueller  As I understood, there is a preview feature on Azure DevOps to use .coverage file to display code coverage in the build results without needing to generate html results separately.

I never ear that. Is true?

**Comment by simonachmueller** [comment]
@gioce90 nope, unfortunately a face news from MS (seems like they tested it, but did not release).

**Comment by skotian2** [comment]
Is this issue resolved with the mentioned steps above ?

**Comment by gioce90** [comment]
@skotian2 is resolved only if you launch a task that use directly the cobertura format, as I described previously. 
If you generate .coverage files (as mentioned in the first comment on this thread) you will not be able to see the report in the "Code coverage" tab (is [a known issue already specified in this thread](https://github.com/microsoft/azure-pipelines-tasks/issues/13632#issuecomment-903107159)):

> If a .coverage file is created and left alone, Azure DevOps will pick it up and use it over anything else you create later. So basically your coverage.xml or any generated HTML report is ignored at the Code Coverage tab.

 But as solution you can:
- collect coverage with coverlet or
-  collect coverage with the native Microsoft code coverage method, but specifying "cobertura" format. I show how [here](https://github.com/microsoft/vstest/issues/981#issuecomment-1034093540) (please add a thumbs-up reaction if you like).

Returning to your question @skotian2 , if this issue is resolved, I see two possible answers:

- if we consider my solutions, yes
- if we think that the real problem is the one mentioned by @elninoisback , i.e if a .coverage file is generated then Azure DevOps ignores the files in .xml format, then no.

**Comment by gioce90** [comment]
> be for Microsoft to update Azure DevOps so that you also get the diff PR integra

I think exactly like you. My solution is just a workaround on a problem that only Microsoft can address.
